### PR TITLE
[CELEBORN-1179] Add support in Celeborn Workers to fetch application meta from the Master

### PR DESCRIPTION
**Pull request by otterc**
### What changes were proposed in this pull request?
This enables a Celeborn Worker to retrieve the application meta from the Master if it hasn't received the secret from the Master before the application attempts to connect to it. Additionally, the Celeborn Worker's SecretRegistry has been converted into an LRU cache to prevent unbounded growth of the registry.

### Why are the changes needed?
This is last change needed for Auth support in Celeborn (https://issues.apache.org/jira/browse/CELEBORN-1011)

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added UTs and part of a bigger change which will be tested end-to-end.

### CODE REVIEW

**Comment by otterc** [comment]
@RexXiong @FMX @pan3793 @mridulm @akpatnam25 @rmcyang @waitinfuture @SteNicholas @melonist Please help review

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/WorkerSecretRegistryImpl.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  private static final Logger LOG = LoggerFactory.getLogger(WorkerSecretRegistryImpl.class);
+
+  // MasterClient is created in Worker after the secret registry is created and this order currently
+  // cannot be changed.
+  // So, we need to set the masterClient after the secret registry is created.
+  private MasterClient masterClient;
+  private final Cache<String, String> secretCache;
+
+  public WorkerSecretRegistryImpl(long maxCacheSize) {
+    secretCache = CacheBuilder.newBuilder().maximumSize(maxCacheSize).build();
```
Comment (SteNicholas) [inline worker/src/main/java/org/apache/celeborn/service/deploy/worker/WorkerSecretRegistryImpl.java:42]: Does `secretCache` need to support expiring?

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 2]
```diff
@@ -4632,4 +4634,12 @@ object CelebornConf extends Logging {
       .version("0.5.0")
       .intConf
       .createWithDefault(8)
+
+  val WORKER_APPLICATION_REGISTRY_CACHE_SIZE: ConfigEntry[Int] =
+    buildConf("celeborn.worker.applicationRegistryCache.size")
```
Comment (SteNicholas) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4639]:
```suggestion
    buildConf("celeborn.worker.applicationRegistry.cache.size")
```

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 3]
```diff
@@ -4632,4 +4634,12 @@ object CelebornConf extends Logging {
...
+      .categories("worker", "auth")
+      .doc("Cache size of the application registry on Workers.")
+      .version("0.5.0")
+      .intConf
+      .createWithDefault(10000)
```
Comment (SteNicholas) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4644]: Why does the default value of cache size is 10000?

File: master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala [hunk 4]
```diff
@@ -1099,6 +1103,26 @@ private[celeborn] class Master(
...
   }
 
+  private def handleRequestForApplicationMeta(
+      context: RpcCallContext,
+      pb: PbApplicationMetaRequest): Unit = {
+    val appId = pb.getAppId
+    logDebug(
+      s"Handling request for application meta info $appId.")
+    if (!secretRegistry.isRegistered(appId)) {
+      logWarning(s"Couldn't find the app .")
```
Comment (SteNicholas) [inline master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala:1113]:
```suggestion
      logWarning(s"Could not find the application meta of $appId.")
```

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/WorkerSecretRegistryImpl.java (hunk 1 above)
Comment (otterc) [reply worker/src/main/java/org/apache/celeborn/service/deploy/worker/WorkerSecretRegistryImpl.java:42]: I don't see a reason to have an expiration on entries. A more complete solution for this is to remove completed applications from the registry for which I created this jira: https://issues.apache.org/jira/browse/CELEBORN-1268. For now, I think just bounding this cache is sufficient.

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala (hunk 3 above)
Comment (otterc) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4644]: I am open to a better default. We have more than 200K apps running per day. The number of apps active at a given will be much smaller than that. I thought 10K is a conservative default and approx will take a few MBs.

File: master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala (hunk 4 above)
Comment (otterc) [reply master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala]: fixed it with other changes.

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala (hunk 2 above)
Comment (otterc) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala]: fixed with other changes.

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/WorkerSecretRegistryImpl.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        LOG.debug(
+            "Successfully fetched the application meta info for " + appId + " from the master");
+        register(pbApplicationMeta.getAppId(), pbApplicationMeta.getSecret());
+        secret = pbApplicationMeta.getSecret();
+      } catch (Throwable e) {
+        // We catch Throwable here because masterClient.askSync declares it in its definition.
+        // If the secret is null, the authentication will fail so just logging the exception here.
+        LOG.error("Failed to fetch the application meta info for {} from the master", appId, e);
+      }
+    }
```
Comment (mridulm) [inline worker/src/main/java/org/apache/celeborn/service/deploy/worker/WorkerSecretRegistryImpl.java:65]: Do we want to move this into a CacheLoader for the cache ?

File: master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala [hunk 6]
```diff
@@ -1099,6 +1103,26 @@ private[celeborn] class Master(
...
+    logDebug(
+      s"Handling request for application meta info $appId.")
+    if (!secretRegistry.isRegistered(appId)) {
+      logWarning(s"Could not find the application meta of $appId.")
+      context.sendFailure(new CelebornException(s"$appId is not registered."))
+    } else {
+      val pbApplicationMeta = PbApplicationMeta.newBuilder()
+        .setAppId(appId)
+        .setSecret(secretRegistry.getSecretKey(appId))
+        .build()
```
[the remark this answers is not in the document]
Comment (otterc) [reply master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala:1119]: Done

File: master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala [hunk 7]
```diff
@@ -1099,6 +1103,26 @@ private[celeborn] class Master(
     }.asJava
   }
 
+  private def handleRequestForApplicationMeta(
+      context: RpcCallContext,
+      pb: PbApplicationMetaRequest): Unit = {
+    val appId = pb.getAppId
+    logDebug(
+      s"Handling request for application meta info $appId.")
```
[the remark this answers is not in the document]
Comment (otterc) [reply master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala]: Done

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/WorkerSecretRegistryImpl.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  @Override
+  public void register(String appId, String secret) {
+    String existingSecret = secretCache.getIfPresent(appId);
+    if (existingSecret != null && !existingSecret.equals(secret)) {
+      throw new IllegalArgumentException(
+          "AppId " + appId + " is already registered. Cannot re-register.");
+    }
+    secretCache.put(appId, secret);
```
Comment (mridulm) [inline worker/src/main/java/org/apache/celeborn/service/deploy/worker/WorkerSecretRegistryImpl.java:102]: super nit: add if `existingSecret != null`

**Review by mridulm — APPROVED** [review]
Looks good to me.
This is the last PR for Authn, right @otterc ?
Awesome to see this feature complete - thanks a lot to the community for the very quick and responsive reviews !!

**Comment by otterc** [comment]
> Looks good to me. This is the last PR for Authn, right @otterc ? Awesome to see this feature complete - thanks a lot to the community for the very quick and responsive reviews !!

Yes, this is the last PR for https://issues.apache.org/jira/browse/CELEBORN-1011 :-). 
Huge thanks to everyone for the quick feedback and the reviews.

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/WorkerSecretRegistryImpl.java [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+  @Override
+  public boolean isRegistered(String appId) {
+    return secretCache.getIfPresent(appId) != null;
+  }
+
+  @Override
+  public void register(String appId, String secret) {
+    String existingSecret = secretCache.getIfPresent(appId);
+    if (existingSecret != null && !existingSecret.equals(secret)) {
+      throw new IllegalArgumentException(
```
[the remark this answers is not in the document]
Comment (otterc) [reply worker/src/main/java/org/apache/celeborn/service/deploy/worker/WorkerSecretRegistryImpl.java:99]: Registration for an application occurs only once, and the associated secret cannot be updated. This approach aligns with the Sasl Authentication mechanism provided by the External Shuffle Service in Spark. Updating the secret is non-trivial and doesn't offer additional benefits in terms of authentication, so it is not supported.

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala (hunk 3 above)
Comment (otterc) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4644]: workerInfo.getApplicationIdSet changes with new apps being added or removed. I am not sure how we could configure the default value of this config to be the size of that set. I did update the code to unregister an app from the worker's secretRegistry from the cleaner thread.

**Pull request closed without merging**